File: web/lib/market/order-book.ts
export type Side="bid"|"ask";
export type Level={price:number;quantity:number};
export class OrderBook{
 bids=new Map<number,number>(); asks=new Map<number,number>(); sequence=0; timestamp=0;
 snapshot(bids:Level[],asks:Level[],timestamp=0,sequence?:number){this.bids=new Map(bids.filter(x=>x.quantity>0).map(x=>[x.price,x.quantity]));this.asks=new Map(asks.filter(x=>x.quantity>0).map(x=>[x.price,x.quantity]));this.advance(timestamp,sequence)}
 update(side:Side,price:number,quantity:number,timestamp=0,sequence?:number){const b=side==="bid"?this.bids:this.asks;if(quantity<=0)b.delete(price);else b.set(price,quantity);this.advance(timestamp,sequence)}
 bestBid(){const x=[...this.bids.entries()].sort((a,b)=>b[0]-a[0])[0];return x?{price:x[0],quantity:x[1]}:null}
 bestAsk(){const x=[...this.asks.entries()].sort((a,b)=>a[0]-b[0])[0];return x?{price:x[0],quantity:x[1]}:null}
 mid(){const b=this.bestBid(),a=this.bestAsk();return b&&a?(b.price+a.price)/2:null}
 spread(){const b=this.bestBid(),a=this.bestAsk();return b&&a?a.price-b.price:null}
 microprice(){const b=this.bestBid(),a=this.bestAsk();if(!b||!a)return null;const t=b.quantity+a.quantity;return t?(a.price*b.quantity+b.price*a.quantity)/t:null}
 depth(levels=10){return{bids:[...this.bids].sort((a,b)=>b[0]-a[0]).slice(0,levels).map(([price,quantity])=>({price,quantity})),asks:[...this.asks].sort((a,b)=>a[0]-b[0]).slice(0,levels).map(([price,quantity])=>({price,quantity}))}}
 imbalance(levels=5){const d=this.depth(levels),b=d.bids.reduce((s,x)=>s+x.quantity,0),a=d.asks.reduce((s,x)=>s+x.quantity,0);return b+a?(b-a)/(b+a):0}
 queueAhead(side:Side,price:number){return Math.max(0,(side==="bid"?this.bids:this.asks).get(price)||0)}
 private advance(timestamp:number,sequence?:number){this.timestamp=timestamp;this.sequence=sequence??this.sequence+1}
}

File: engine/market/order_book.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable


@dataclass
class BookLevel:
    price: float
    quantity: float


@dataclass
class OrderBook:
    """Deterministic L2 order-book state used by replay and simulation."""

    bids: dict[float, float] = field(default_factory=dict)
    asks: dict[float, float] = field(default_factory=dict)
    sequence: int = 0
    timestamp_ns: int = 0

    def apply_snapshot(
        self,
        bids: Iterable[tuple[float, float]],
        asks: Iterable[tuple[float, float]],
        *,
        timestamp_ns: int = 0,
        sequence: int | None = None,
    ) -> None:
        self.bids = {float(p): float(q) for p, q in bids if q > 0}
        self.asks = {float(p): float(q) for p, q in asks if q > 0}
        self._advance(timestamp_ns, sequence)

    def update(
        self,
        side: str,
        price: float,
        quantity: float,
        *,
        timestamp_ns: int = 0,
        sequence: int | None = None,
    ) -> None:
        book = self.bids if side.lower() == "bid" else self.asks
        price = float(price)
        quantity = float(quantity)
        if quantity <= 0:
            book.pop(price, None)
        else:
            book[price] = quantity
        self._advance(timestamp_ns, sequence)

    def best_bid(self) -> BookLevel | None:
        return self._best(self.bids, reverse=True)

    def best_ask(self) -> BookLevel | None:
        return self._best(self.asks, reverse=False)

    def mid_price(self) -> float | None:
        bid, ask = self.best_bid(), self.best_ask()
        if not bid or not ask:
            return None
        return (bid.price + ask.price) / 2.0

    def spread(self) -> float | None:
        bid, ask = self.best_bid(), self.best_ask()
        if not bid or not ask:
            return None
        return ask.price - bid.price

    def microprice(self) -> float | None:
        bid, ask = self.best_bid(), self.best_ask()
        if not bid or not ask:
            return None
        total = bid.quantity + ask.quantity
        if total <= 0:
            return None
        return (ask.price * bid.quantity + bid.price * ask.quantity) / total

    def depth(self, levels: int = 10) -> dict[str, list[BookLevel]]:
        return {
            "bids": [BookLevel(p, self.bids[p]) for p in sorted(self.bids, reverse=True)[:levels]],
            "asks": [BookLevel(p, self.asks[p]) for p in sorted(self.asks)[:levels]],
        }

    def imbalance(self, levels: int = 5) -> float:
        d = self.depth(levels)
        bid_qty = sum(x.quantity for x in d["bids"])
        ask_qty = sum(x.quantity for x in d["asks"])
        total = bid_qty + ask_qty
        return 0.0 if total <= 0 else (bid_qty - ask_qty) / total

    def queue_ahead(self, side: str, price: float) -> float:
        """Approximate displayed quantity ahead at the order's price level."""
        book = self.bids if side.lower() == "buy" else self.asks
        return max(0.0, float(book.get(float(price), 0.0)))

    def snapshot(self, levels: int = 10) -> dict:
        d = self.depth(levels)
        return {
            "timestamp_ns": self.timestamp_ns,
            "sequence": self.sequence,
            "best_bid": d["bids"][0].__dict__ if d["bids"] else None,
            "best_ask": d["asks"][0].__dict__ if d["asks"] else None,
            "mid": self.mid_price(),
            "spread": self.spread(),
            "microprice": self.microprice(),
            "imbalance": self.imbalance(min(levels, 5)),
            "bids": [x.__dict__ for x in d["bids"]],
            "asks": [x.__dict__ for x in d["asks"]],
        }

    def _advance(self, timestamp_ns: int, sequence: int | None) -> None:
        self.timestamp_ns = int(timestamp_ns)
        if sequence is None:
            self.sequence += 1
        else:
            self.sequence = int(sequence)

    @staticmethod
    def _best(book: dict[float, float], reverse: bool) -> BookLevel | None:
        if not book:
            return None
        price = sorted(book, reverse=reverse)[0]
        return BookLevel(price, book[price])
